connect/zypper_test.go

```diff
@@ -1,20 +1,66 @@
 package connect
 
 import (
+	"fmt"
+	"os"
+	"os/exec"
 	"testing"
 )
 
-var testProducts = `<?xml version='1.0'?>
-<stream>
-  <product-list>
-    <product name="SUSE-MicroOS" version="5.0" release="1" epoch="0" arch="x86_64" vendor="SUSE" summary="SUSE Linux Enterprise Micro 5.0" repo="@System" productline="SUSE-MicroOS" registerrelease="" shortname="SUSE Linux Enterprise Micro" flavor="" isbase="true" installed="true"><endoflife time_t="1606694400" text="2020-11-30T00:00:00Z"/><registerflavor/><description>SUSE Linux Enterprise Micro 5.0</description></product>
-    <product name="suse-openstack-cloud" version="8" release="0" epoch="0" arch="x86_64" vendor="SUSE LINUX GmbH, Nuernberg, Germany" summary="SUSE OpenStack Cloud 8" repo="@System" productline="suse-openstack-cloud" registerrelease="" shortname="SOC8" flavor="POOL" isbase="false" installed="true"><endoflife time_t="1622419200" text="2021-05-31T00:00:00+00"/><registerflavor>extension</registerflavor><description>SUSE OpenStack Cloud 8</description></product>
-  </product-list>
-</stream>
-`
+const (
+	testCaseInstalledProducts       = "testCaseInstalledProducts"
+	testCaseInstalledProductsNoBase = "testCaseInstalledProductsNoBase"
+)
+
+var zypperExecTestCase = ""
+
+func zypperExecMock(command string, args ...string) *exec.Cmd {
+	cs := []string{"-test.run=TestZypperHelper", "--", command}
+	cs = append(cs, args...)
+	cmd := exec.Command(os.Args[0], cs...)
+	cmd.Env = os.Environ()
+	cmd.Env = append(cmd.Env, "GO_WANT_HELPER_PROCESS=1")
+	cmd.Env = append(cmd.Env, "GO_TESTCASE="+zypperExecTestCase)
+	return cmd
+}
+
+// This is not a real test. It's used for mocking output of commands.
+func TestZypperHelper(t *testing.T) {
+	// if ran directly, do nothing
+	if os.Getenv("GO_WANT_HELPER_PROCESS") != "1" {
+		return
+	}
+	testcase := os.Getenv("GO_TESTCASE")
+	switch testcase {
+	case testCaseInstalledProducts:
+		fmt.Fprintf(os.Stdout, string(readTestFile("products.xml", t)))
+		os.Exit(0)
+	case testCaseInstalledProductsNoBase:
+		fmt.Fprintf(os.Stdout, string(readTestFile("products-no-base.xml", t)))
```

Comment:
you don't need to use the string() if you
`fmt.Fprintf(os.Stdout, "%s", readTestFile("products-no-base.xml", t))`
but it's all the same

Reply:
I changed it to `os.Stdout.Write(readTestFile("products-no-base.xml", t))`.
I think this is the best pattern in case we need some binary test files at some point.

Reply:
yep that's even better